# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/syedwaqasahmad/FlyRank-ML-Internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

I'm choosing a freestyle direction inspired by churn analysis: Content Visibility Churn. Instead of customers leaving a subscription, I'm treating a page's decline in search visibility as a "churn" event — a page that was performing well and then drops off. This reframes the same churn-prediction problem (who's at risk of leaving, and why) onto FlyRank's actual content data, using the trend_direction field as the churn signal.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [19]:
import os, subprocess
REPO_URL = "https://github.com/syedwaqasahmad/FlyRank-ML-Internship"
REPO_DIR = "FlyRank-ML-Internship"
if not os.path.isdir(REPO_DIR):
    subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
os.chdir(REPO_DIR)

import pandas as pd
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")
print(df.shape[0], "pages,", df.shape[1], "columns")

30000 pages, 44 columns


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

Decision this improves: which pages are at highest risk of "churning" (losing search visibility) so a reviewer can intervene before it happens, not after.

Who acts on it: a content/SEO reviewer with limited time, deciding which pages to check first.

Cost of a wrong call: a false positive wastes review time on a page that was never at risk; a false negative means a page silently churns out of visibility and the reviewer finds out too late to act.

In [20]:
# Confirming the churn signal we're using is real and present in the data
print("Churn signal used:", "trend_direction == 'down'")
print("Values in trend_direction:", df["trend_direction"].unique())

Churn signal used: trend_direction == 'down'
Values in trend_direction: ['down' 'stable' 'new' 'up' 'flat']


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

Using the real FlyRank starter data, I looked at how many pages have actually "churned" (declined in search visibility), and how many show the early warning signs of churn risk (stale + still getting traffic) before checking whether those warning signs actually line up with real declines.

In [21]:
churned = df["trend_direction"].str.lower().eq("down")
at_risk = (df["days_since_last_update"] >= 180) & (df["impressions_90d"] >= 100)

print(f"Pages that have 'churned' (declining): {churned.sum()} of {len(df)} ({churned.mean()*100:.1f}%)")
print(f"Pages stale + still getting traffic (churn-risk candidates): {at_risk.sum()} of {len(df)} ({at_risk.mean()*100:.1f}%)")
print(f"Of the churn-risk candidates, how many actually churned: {(at_risk & churned).sum()} of {at_risk.sum()}")

Pages that have 'churned' (declining): 16262 of 30000 (54.2%)
Pages stale + still getting traffic (churn-risk candidates): 35 of 30000 (0.1%)
Of the churn-risk candidates, how many actually churned: 26 of 35


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

What I can claim: this is an observed, directional signal — pages that are stale and still getting search visibility "churn" (decline) at a noticeably higher rate (74%) than the overall base rate (54.2%). That's evidence worth building on, not proof of a mechanism.

What I can't claim: this doesn't prove causation — I can't say staleness causes decline, only that they're correlated in this data. I'm also not predicting Google's algorithm; this supports a human reviewer's prioritization decision, it doesn't replace their judgment.

In [22]:
# Quick sanity check that this notebook's own numbers are internally consistent
assert (at_risk & churned).sum() <= at_risk.sum()
print("Numbers check out.")

Numbers check out.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.